README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu/check-disk-usage)
![Go Test](https://github.com/sensu/check-disk-usage/workflows/Go%20Test/badge.svg)
![goreleaser](https://github.com/sensu/check-disk-usage/workflows/goreleaser/badge.svg)

# Sensu disk usage check

## Table of Contents
- [Overview](#overview)
- [Usage](#usage)
  - [Help output](#help-output)
  - [Usage notes](#usage-notes)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Check definition](#check-definition)
- [Installation from source](#installation-from-source)
- [Contributing](#contributing)

## Overview

The Sensu disk usage check is a [Sensu Check][2] that reports on disk usage
allowing for the inclusion or exclusion of certain file systems and/or file
system types.

## Usage

### Help output
```
Cross platform disk usage check for Sensu

Usage:
  check-disk-usage [flags]
  check-disk-usage [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -i, --include-fs-type strings   Comma separated list of file system types to check
  -e, --exclude-fs-type strings   Comma separated list of file system types to exclude from checking
  -I, --include-fs-path strings   Comma separated list of file system paths to check
  -E, --exclude-fs-path strings   Comma separated list of file system paths to exclude from checking
  -w, --warning float             Warning threshold for file system usage (default 85)
  -c, --critical float            Critical threshold for file system usage (default 95)
  -p, --include-pseudo-fs         Include pseudo-filesystems (e.g. tmpfs) (default false)
  -r, --include-read-only         Include read-only filesystems (default false)
  -f, --fail-on-error             Fail and exit on errors getting file system usage (e.g. permission denied) (default false)
  -m, --magic float               Magic factor to adjust warn/crit thresholds. Example: .9 (default 1)
  -l, --minimum float             Minimum size to adjust (in GiB) (default 100)
  -n, --normal float              Value in GiB. Levels are not adapted for filesystems of exactly this size, where levels are reduced for smaller filesystems and raised for larger filesystems. (default 20)
  -H, --human-readable            print sizes in powers of 1024 (default false)
      --metrics                   Output metrics instead of human readable output
      --tags strings              Comma separated list of additional metrics tags using key=value format.
  -h, --help                      help for check-disk-usage

Use "check-disk-usage [command] --help" for more information about a command.
```

### Usage notes

* The include and exclude options for both file system type and path are
mutually exclusive (e.g. you can not use `--exclude-fs-type` and
`--include-fs-type` on the same check).
* The file system path on Linux/Unix/macOS systems means the file system mount
point (e.g. /, /tmp, /home)
* The file system path on Windows refers to the drive letter (e.g. C:, D:).
Volumes mounted via UNC paths are not checked.
* File system types and paths on Windows are capitalized and need to be
specified as such (e.g. NTFS, C:)
* The `--include-pseudo-fs` option is false by default meaning that on Linux
systems file system with types such as tmpfs (e.g. /dev, /run, etc.) will
be ignored. This takes precedence over any explicit includes or excludes.
* The `--include-read-only` checks for the `ro` mount option on Linux and the
`read-only` mount option on macOS and read-only volumes on Windows.  By default
these file systems will be ignored. The rationale being that if a monitored
system does not have write access to a file system, it cannot be used to create
files (not the source of the problem) nor can it be used to clean up the file
system (not a part of the solution).
* The `--fail-on-error` option determines what occurs if the check encounters an
error, such as `permission denied` for a file system.  If true, the check will
exit with as a critical failure and provide the error message.  If false (the
defaut), it will specify unknown for that file system, provide the error and
continue to check the remaining file systems as expected.
* The `--human-readable` (False by default) option determines if you prefer
to display sizes of different drives in a human format. (Like df Unix/linux
command.)
* The `--magic`, `--normal`, and `--minimum` options work together to adjust 
thresholds for larger filesystems. When a filesystem is larger than `--minimum` 
the `--magic` scaling factor is used. For filesystems larger than `--normal` a magic scaling 
factor less than 1 increases warning and critical thresholds. On filesystems
smaller than `--normal`, a magic scaling factor less than 1 descreases thesholds.
The idea being there's a bit more/less margin available for filesystems 
larger/smaller relative to `--normal` based on similar data accomulation patterns, so the magic scaling factor lets you set a `normal` threshold and adjust the warning and critical thresholds up or down using a calculation.  The default `--magic` value, 1.0, will not adapt threshold percentages for volumes. 

#### Magic Scaling Calculation

```
normalized_disk_size = disk_size_GiB/normal_size_GiB
rescaled_disk_size = (normalized_disk_size)^magic_scale_factor
scale = rescaled_disk_size / normalized_disk_size
new_percent_threhold = 100.0 - ((100.0 - original_percent_threshold) * scale)
```

* For disks larger than `--normal` and `--magic` values less than 1, 
result in `scale < 1` and `new_percent_threshold > original_percent_threshold`  
* For disks smaller than `--normal`and `--magic` values less than 1 result in `scale > 1` 
result in `scale > 1` and `new_percent_threshold < original_percent_threshold`  

## Configuration

### Asset registration

[Sensu Assets][4] are the best way to make use of this plugin. If you're not
using an asset, please consider doing so! If you're using sensuctl 5.13 with
Sensu Backend 5.13 or later, you can use the following command to add the asset:

```
sensuctl asset add sensu/check-disk-usage
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index][3].

### Check definition

#### Linux example

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: check-disk-usage
  namespace: default
spec:
  command: >-
    check-disk-usage
    --include-fs-type "xfs,ext4"
    --exclude-fs-path "/boot"
    --warning 90
    --critical 95
  subscriptions:
  - system
  runtime_assets:
  - sensu/check-disk-usage
```

#### Windows example
```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: check-disk-usage
  namespace: default
spec:
  command: >-
    check-disk-usage
    --include-fs-type "NTFS"
    --exclude-fs-path "C:,D:"
    --warning 90
    --critical 95
  subscriptions:
  - system
  runtime_assets:
  - sensu/check-disk-usage
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an
Asset. If you would like to compile and install the plugin from source or
contribute to it, download the latest version or create an executable from this
source.

From the local path of the check-disk-usage repository:

```
go build
```

## Contributing

For more information about contributing to this plugin, see [Contributing][1].

[1]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[2]: https://docs.sensu.io/sensu-go/latest/reference/checks/
[3]: https://bonsai.sensu.io/assets/sensu/check-disk-usage
[4]: https://docs.sensu.io/sensu-go/latest/reference/assets/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
### Added
- Implement magic scaling factor for large filesystems similar to original ruby plugin.

### Fixed
- corrected armv7 asset build filter in .bonsai.yml. Thanks @DJM0.

## [0.7.0] - 2022-04-06

### Fixed
- Timestamp precision from second to millisecond

## [0.5.0] - 2021-10-18

### Added
- Added --metrics option to change output into something Sensu agent can register as tagged metrics
- Added --tags option to make it possible to add additional tags for all metrics
- Added --human-readable option to support ignoring human-readable option like df unix/linux command

## [0.4.2] - 2021-03-31

### Changed
- Fixed typo(s)
- README update(s)

## [0.4.1] - 2021-02-14

### Changed
- Fixed README re: read-only

## [0.4.0] - 2021-02-14

### Added
- Added --include-read-only option to support ignoring read-only file systems

## [0.3.1] - 2020-12-30

### Changed
- Added newline to end of UNKNOWN error output

## [0.3.0] - 2020-12-30

### Changed
- Moved included/excluded checks against partition information up before querying usage

## [0.2.0] - 2020-12-30

### Added
- Added flags --include-pseudo-fs and --fail-on-error

### Changed
- Changed the behavior for failures to get fs usage based on flag
- Changed call to get partitions to use --include-pseudo-fs flag

## [0.1.1] - 2020-12-29

### Change
- Added appropriate error checking

## [0.1.0] - 2020-12-14

### Added
- Initial release


main_test.go
package main

import (
	"testing"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-plugin-sdk/sensu"
	"github.com/stretchr/testify/assert"
)

func clearOptions() {
	plugin.IncludeFSType = nil
	plugin.ExcludeFSType = nil
	plugin.IncludeFSPath = nil
	plugin.ExcludeFSPath = nil
	plugin.ExtraTags = nil
	plugin.Warning = 85.0
	plugin.Critical = 95.0
	plugin.MinimumGiB = 100.0
	plugin.NormalGiB = 20.0
	plugin.MetricsMode = false
	plugin.Verbose = false
}

func TestAdjPercent(t *testing.T) {
	clearOptions()
	percent := 80.0
	assert := assert.New(t)
	result := adjPercent(0.0, percent)
	assert.Equal(percent, result)
	plugin.NormalGiB = 0.0
	result = adjPercent(100.0, percent)
	assert.Equal(percent, result)
	plugin.NormalGiB = 20.0
	result = adjPercent(100.0, -1.0)
	assert.Equal(0.0, result)
	plugin.Verbose = true
	result = adjPercent(0.8*plugin.NormalGiB*1.074e+9, percent)
	assert.Greater(percent, result)
	assert.LessOrEqual(0.0, result)
	assert.GreaterOrEqual(100.0, result)
	result = adjPercent(10.0*plugin.NormalGiB*1.074e+9, percent)
	assert.Less(percent, result)
	assert.LessOrEqual(0.0, result)
	assert.GreaterOrEqual(100.0, result)

}

func TestIsValidFSType(t *testing.T) {
	clearOptions()
	assert := assert.New(t)
	plugin.IncludeFSType = []string{"ext4", "xfs"}
	result := isValidFSType("ext4")
	assert.True(result)
	result = isValidFSType("xfs")
	assert.True(result)
	result = isValidFSType("bad")
	assert.False(result)
	clearOptions()
	plugin.ExcludeFSType = []string{"ext4", "xfs"}
	result = isValidFSType("bad")
	assert.True(result)
	result = isValidFSType("ext4")
	assert.False(result)
	result = isValidFSType("xfs")
	assert.False(result)

}

func TestIsValidFSPath(t *testing.T) {
	clearOptions()
	assert := assert.New(t)
	result := isValidFSPath("/mount1")
	assert.True(result)
	plugin.IncludeFSPath = []string{"/mount1/", "/mount2"}
	result = isValidFSPath("/mount1/")
	assert.True(result)
	result = isValidFSPath("/mount2")
	assert.True(result)
	result = isValidFSPath("/mount3")
	assert.False(result)
	clearOptions()
	plugin.ExcludeFSPath = []string{"/mount1", "/mount2"}
	result = isValidFSPath("/mount3")
	assert.True(result)
	result = isValidFSPath("/mount1")
	assert.False(result)
	result = isValidFSPath("/mount2")
	assert.False(result)

}

func TestCheckArgs(t *testing.T) {
	clearOptions()
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	plugin.IncludeFSType = []string{"ext4", "xfs}"}
	plugin.ExcludeFSType = []string{"tmpfs", "devtmpfs"}
	i, e := checkArgs(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.Error(e)
	plugin.ExcludeFSType = []string{}
	plugin.IncludeFSPath = []string{"/", "/home"}
	plugin.ExcludeFSPath = []string{"/tmp"}
	i, e = checkArgs(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.Error(e)
	plugin.ExcludeFSPath = []string{}
	plugin.Warning = float64(80)
	plugin.Critical = float64(70)
	i, e = checkArgs(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.Error(e)
	plugin.Critical = float64(90)
	i, e = checkArgs(event)
	assert.Equal(sensu.CheckStateOK, i)
	assert.NoError(e)
	plugin.ExtraTags = []string{"tmpfs", "devtmpfs"}
	i, e = checkArgs(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.Error(e)
	plugin.ExtraTags = []string{"key1=val1", "key2=val2"}
	i, e = checkArgs(event)
	assert.Equal(sensu.CheckStateOK, i)
	assert.NoError(e)

}

func TestExecuteCheck(t *testing.T) {
	clearOptions()
	plugin.MetricsMode = true
	plugin.ExcludeFSPath = []string{"/dev"}
	assert := assert.New(t)
	event := corev2.FixtureEvent("entity1", "check1")
	i, e := executeCheck(event)
	assert.Equal(sensu.CheckStateOK, i)
	assert.NoError(e)
	clearOptions()
	plugin.Warning = -200.0
	plugin.Critical = -100.0
	// Set minimum extremely high to disable magic adjustment
	plugin.MinimumGiB = 100000.0
	i, e = executeCheck(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.NoError(e)
	// Set minimum low and normal high to enable magic adjustment
	plugin.NormalGiB = 100000.0
	plugin.MinimumGiB = 1.0
	plugin.Verbose = true
	i, e = executeCheck(event)
	assert.Equal(sensu.CheckStateCritical, i)
	assert.NoError(e)
}


main.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	human "github.com/dustin/go-humanize"
	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-plugin-sdk/sensu"
	"github.com/shirou/gopsutil/disk"
)

// Config represents the check plugin config.
type Config struct {
	sensu.PluginConfig
	IncludeFSType   []string
	ExcludeFSType   []string
	IncludeFSPath   []string
	ExcludeFSPath   []string
	ExtraTags       []string
	Warning         float64
	Critical        float64
	NormalGiB       float64
	Magic           float64
	MinimumGiB      float64
	IncludePseudo   bool
	IncludeReadOnly bool
	FailOnError     bool
	HumanReadable   bool
	MetricsMode     bool
	Verbose         bool
}

type MetricGroup struct {
	Comment string
	Type    string
	Name    string
	Metrics []Metric
}

func (g *MetricGroup) AddMetric(tags map[string]string, value float64, timeNow int64) {
	g.Metrics = append(g.Metrics, Metric{
		Tags:      tags,
		Timestamp: timeNow,
		Value:     value,
	})
}
func (g *MetricGroup) Output() {
	var output string
	metricName := strings.Replace(g.Name, ".", "_", -1)
	fmt.Printf("# HELP %s [%s] %s\n", metricName, g.Type, g.Comment)
	fmt.Printf("# TYPE %s %s\n", metricName, g.Type)
	for _, m := range g.Metrics {
		tagStr := ""
		for tag, tvalue := range m.Tags {
			if len(tagStr) > 0 {
				tagStr = tagStr + ","
			}
			tagStr = tagStr + tag + "=\"" + tvalue + "\""
		}
		if len(tagStr) > 0 {
			tagStr = "{" + tagStr + "}"
		}
		output = strings.Join(
			[]string{metricName + tagStr, fmt.Sprintf("%v", m.Value), strconv.FormatInt(m.Timestamp, 10)}, " ")
		fmt.Println(output)
	}
	fmt.Println("")
}

type Metric struct {
	Tags      map[string]string
	Timestamp int64
	Value     float64
}

var (
	tags      = map[string]string{}
	extraTags = map[string]string{}
	plugin    = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "check-disk-usage",
			Short:    "Cross platform disk usage check for Sensu",
			Keyspace: "sensu.io/plugins/check-disk-usage/config",
		},
	}

	options = []sensu.ConfigOption{
		&sensu.SlicePluginConfigOption[string]{
			Path:      "include-fs-type",
			Env:       "",
			Argument:  "include-fs-type",
			Shorthand: "i",
			Default:   []string{},
			Usage:     "Comma separated list of file system types to check",
			Value:     &plugin.IncludeFSType,
		},
		&sensu.SlicePluginConfigOption[string]{
			Path:      "exclude-fs-type",
			Env:       "",
			Argument:  "exclude-fs-type",
			Shorthand: "e",
			Default:   []string{},
			Usage:     "Comma separated list of file system types to exclude from checking",
			Value:     &plugin.ExcludeFSType,
		},
		&sensu.SlicePluginConfigOption[string]{
			Path:      "include-fs-path",
			Env:       "",
			Argument:  "include-fs-path",
			Shorthand: "I",
			Default:   []string{},
			Usage:     "Comma separated list of file system paths to check",
			Value:     &plugin.IncludeFSPath,
		},
		&sensu.SlicePluginConfigOption[string]{
			Path:      "exclude-fs-path",
			Env:       "",
			Argument:  "exclude-fs-path",
			Shorthand: "E",
			Default:   []string{},
			Usage:     "Comma separated list of file system paths to exclude from checking",
			Value:     &plugin.ExcludeFSPath,
		},
		&sensu.PluginConfigOption[float64]{
			Path:      "warning",
			Env:       "",
			Argument:  "warning",
			Shorthand: "w",
			Default:   float64(85),
			Usage:     "Warning threshold for file system usage",
			Value:     &plugin.Warning,
		},
		&sensu.PluginConfigOption[float64]{
			Path:      "critical",
			Env:       "",
			Argument:  "critical",
			Shorthand: "c",
			Default:   float64(95),
			Usage:     "Critical threshold for file system usage",
			Value:     &plugin.Critical,
		},
		&sensu.PluginConfigOption[float64]{
			Path:      "normal",
			Env:       "",
			Argument:  "normal",
			Shorthand: "n",
			Default:   float64(20),
			Usage: `Value in GiB. Levels are not adapted for filesystems of exactly this ` +
				`size, where levels are reduced for smaller filesystems and raised ` +
				`for larger filesystems.`,
			Value: &plugin.NormalGiB,
		},
		&sensu.PluginConfigOption[float64]{
			Path:      "magic",
			Env:       "",
			Argument:  "magic",
			Shorthand: "m",
			Default:   float64(1),
			Usage:     `Magic factor to adjust warn/crit thresholds. Example: .9`,
			Value:     &plugin.Magic,
		},
		&sensu.PluginConfigOption[float64]{
			Path:      "minimum",
			Env:       "",
			Argument:  "minimum",
			Shorthand: "l",
			Default:   float64(100),
			Usage:     `Minimum size to adjust (in GiB)`,
			Value:     &plugin.MinimumGiB,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "include-pseudo-fs",
			Env:       "",
			Argument:  "include-pseudo-fs",
			Shorthand: "p",
			Default:   false,
			Usage:     "Include pseudo-filesystems (e.g. tmpfs) (default false)",
			Value:     &plugin.IncludePseudo,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "fail-on-error",
			Env:       "",
			Argument:  "fail-on-error",
			Shorthand: "f",
			Default:   false,
			Usage:     "Fail and exit on errors getting file system usage (e.g. permission denied) (default false)",
			Value:     &plugin.FailOnError,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "include-read-only",
			Env:       "",
			Argument:  "include-read-only",
			Shorthand: "r",
			Default:   false,
			Usage:     "Include read-only filesystems (default false)",
			Value:     &plugin.IncludeReadOnly,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "human-readable",
			Env:       "",
			Argument:  "human-readable",
			Shorthand: "H",
			Default:   false,
			Usage:     "print sizes in powers of 1024 (default false)",
			Value:     &plugin.HumanReadable,
		},
		&sensu.PluginConfigOption[bool]{
			Path:     "metrics",
			Env:      "",
			Argument: "metrics",
			Default:  false,
			Usage:    "Output metrics instead of human readable output",
			Value:    &plugin.MetricsMode,
		},
		&sensu.SlicePluginConfigOption[string]{
			Path:     "tags",
			Env:      "",
			Argument: "tags",
			Default:  []string{},
			Usage:    "Comma separated list of additional metrics tags using key=value format.",
			Value:    &plugin.ExtraTags,
		},
	}
)

func main() {
	check := sensu.NewGoCheck(&plugin.PluginConfig, options, checkArgs, executeCheck, false)
	check.Execute()
}

func checkArgs(event *corev2.Event) (int, error) {
	if len(plugin.IncludeFSType) > 0 && len(plugin.ExcludeFSType) > 0 {
		return sensu.CheckStateCritical, fmt.Errorf("--include-fs-type and --exclude-fs-type are mutually exclusive")
	}
	if len(plugin.IncludeFSPath) > 0 && len(plugin.ExcludeFSPath) > 0 {
		return sensu.CheckStateCritical, fmt.Errorf("--include-fs-path and --exclude-fs-path are mutually exclusive")
	}
	if plugin.Warning >= plugin.Critical {
		return sensu.CheckStateCritical, fmt.Errorf("--warning value can not be greater than or equal to --critical value")
	}
	for _, tagString := range plugin.ExtraTags {
		fmt.Println(tagString)
		parts := strings.Split(tagString, `=`)
		if len(parts) == 2 {
			extraTags[parts[0]] = parts[1]
		} else {
			return sensu.CheckStateCritical, fmt.Errorf("Failed to parse input tag: %s", tagString)
		}
	}
	return sensu.CheckStateOK, nil
}

func executeCheck(event *corev2.Event) (int, error) {
	var (
		criticals int
		warnings  int
	)

	timeNow := time.Now().UnixNano() / 1000000
	parts, err := disk.Partitions(plugin.IncludePseudo)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("Failed to get partitions, error: %v", err)
	}

	metricGroups := map[string]*MetricGroup{
		"disk.critical": &MetricGroup{
			Name:    "disk.critical",
			Type:    "GAUGE",
			Comment: "non-zero value indicates mountpoint usage is above critical threshold",
			Metrics: []Metric{},
		},
		"disk.warning": &MetricGroup{
			Name:    "disk.warning",
			Type:    "GAUGE",
			Comment: "non-zero value indicates mountpoint usage is above warning threshold",
			Metrics: []Metric{},
		},
		"disk.percent_used": &MetricGroup{
			Name:    "disk.percent_usage",
			Type:    "GAUGE",
			Comment: "Percentage of mounted volume used",
			Metrics: []Metric{},
		},
		"disk.total_bytes": &MetricGroup{
			Name:    "disk.total_bytes",
			Type:    "GAUGE",
			Comment: "Total size in bytes of mounted volumed",
			Metrics: []Metric{},
		},
		"disk.used_bytes": &MetricGroup{
			Name:    "disk.used_bytes",
			Type:    "GAUGE",
			Comment: "Used size in bytes of mounted volumed",
			Metrics: []Metric{},
		},
		"disk.free_bytes": &MetricGroup{
			Name:    "disk.free_bytes",
			Type:    "GAUGE",
			Comment: "Free size in bytes of mounted volumed",
			Metrics: []Metric{},
		},
	}

	for _, p := range parts {
		tags = map[string]string{}
		for key, value := range extraTags {
			tags[key] = value
		}
		// Ignore excluded (or non-included) file system types
		if !isValidFSType(p.Fstype) {
			continue
		}

		// Ignore excluded (or non-included) file systems
		if !isValidFSPath(p.Mountpoint) {
			continue
		}

		// Ignore read-only file systems?
		if !plugin.IncludeReadOnly && isReadOnly(p.Opts) {
			continue
		}

		tags["mountpoint"] = p.Mountpoint
		device := p.Mountpoint
		if len(device) == 0 {
			continue
		}
		s, err := disk.Usage(device)
		if err != nil {
			if plugin.FailOnError {
				return sensu.CheckStateCritical, fmt.Errorf("Failed to get disk usage for %s, error: %v", device, err)
			}
			if !plugin.MetricsMode {
				fmt.Printf("%s  UNKNOWN: dev: <%s> - error: %v\n", plugin.PluginConfig.Name, device, err)
			}
			continue
		}

		// Ignore empty file systems
		if s.Total == 0 {
			continue
		}

		// implement magic factor for larger file systems?
		tot := float64(s.Total)
		bcrit := plugin.Critical
		bwarn := plugin.Warning
		if tot > (plugin.MinimumGiB * math.Pow(1024, 3)) {
			bcrit = adjPercent(tot, plugin.Critical)
			bwarn = adjPercent(tot, plugin.Warning)
		}

		crit := 0
		warn := 0
		if plugin.Verbose {
			fmt.Printf("tot: %v  used: %v crit: %v  warn: %v bcrit: %v bwarn: %v\n", tot, s.UsedPercent, plugin.Critical, plugin.Warning, bcrit, bwarn)
		}
		if s.UsedPercent >= bcrit {
			criticals++
			crit = 1
		} else if s.UsedPercent >= bwarn {
			warnings++
			warn = 1
		}
		metricGroups["disk.critical"].AddMetric(tags, float64(crit), timeNow)
		metricGroups["disk.warning"].AddMetric(tags, float64(warn), timeNow)
		if !plugin.MetricsMode {
			fmt.Printf("%s ", plugin.PluginConfig.Name)
			if crit > 0 {
				fmt.Printf("CRITICAL: ")
			} else if warn > 0 {
				fmt.Printf(" WARNING: ")
			} else {
				fmt.Printf("      OK: ")
			}
			if plugin.HumanReadable {
				fmt.Printf("%s %.2f%% - Total: %s, Used: %s, Free: %s\n",
					p.Mountpoint, s.UsedPercent, human.IBytes(s.Total), human.IBytes(s.Used), human.IBytes(s.Free))
			} else {
				fmt.Printf("%s %.2f%% - Total: %s, Used: %s, Free: %s\n",
					p.Mountpoint, s.UsedPercent, human.Bytes(s.Total), human.Bytes(s.Used), human.Bytes(s.Free))
			}
		}
		metricGroups["disk.percent_used"].AddMetric(tags, float64(s.UsedPercent), timeNow)
		metricGroups["disk.total_bytes"].AddMetric(tags, float64(s.Total), timeNow)
		metricGroups["disk.used_bytes"].AddMetric(tags, float64(s.Used), timeNow)
		metricGroups["disk.free_bytes"].AddMetric(tags, float64(s.Free), timeNow)
	}
	tags = map[string]string{}
	for key, value := range extraTags {
		tags[key] = value
	}
	tags["mountpoint"] = "any"
	anyCritical := func() float64 {
		if criticals > 0 {
			return 1
		} else {
			return 0
		}
	}()
	metricGroups["disk.critical"].AddMetric(tags, anyCritical, timeNow)
	anyWarning := func() float64 {
		if warnings > 0 {
			return 1
		} else {
			return 0
		}
	}()
	metricGroups["disk.warning"].AddMetric(tags, anyWarning, timeNow)
	if plugin.MetricsMode {
		// output metrics in a fixed order
		metricGroups["disk.critical"].Output()
		metricGroups["disk.warning"].Output()
		metricGroups["disk.percent_used"].Output()
		metricGroups["disk.total_bytes"].Output()
		metricGroups["disk.used_bytes"].Output()
		metricGroups["disk.free_bytes"].Output()
	}
	if criticals > 0 {
		return sensu.CheckStateCritical, nil
	} else if warnings > 0 {
		return sensu.CheckStateWarning, nil
	}

	return sensu.CheckStateOK, nil
}

func isValidFSType(fsType string) bool {
	// should i account for case insensitive searches for windows (ntfs vs. NTFS)?
	if len(plugin.IncludeFSType) > 0 && contains(plugin.IncludeFSType, fsType) {
		return true
	} else if len(plugin.IncludeFSType) > 0 {
		return false
	} else if len(plugin.ExcludeFSType) > 0 && contains(plugin.ExcludeFSType, fsType) {
		return false
	}

	// either not in exclude list or neither list is specified
	return true
}

func isValidFSPath(fsPath string) bool {
	// should i account for case insensitive searches for windows (c: vs. C:)?
	if len(plugin.IncludeFSPath) > 0 && contains(plugin.IncludeFSPath, fsPath) {
		return true
	} else if len(plugin.IncludeFSPath) > 0 {
		return false
	} else if len(plugin.ExcludeFSPath) > 0 && contains(plugin.ExcludeFSPath, fsPath) {
		return false
	}

	// either not in exclude list or neither list is specified
	return true
}

func isReadOnly(mountOpts string) bool {
	mOpts := strings.Split(mountOpts, ",")
	// "ro" should cover Linux, macOS, and Windows, "read-only" is reportd by mount(8)
	// on macOS so check for it, just in case
	if contains(mOpts, "ro") || contains(mOpts, "read-only") {
		return true
	}
	return false
}

func contains(a []string, s string) bool {
	for _, v := range a {
		if v == s {
			return true
		}
	}
	return false
}

func adjPercent(sizeInBytes float64, percent float64) float64 {
	if sizeInBytes <= 0.0 {
		return percent
	}
	if plugin.NormalGiB <= 0.0 {
		return percent
	}
	if percent <= 0.0 {
		return 0.0
	}
	hsize := (sizeInBytes / math.Pow(1024.0, 3)) / plugin.NormalGiB
	felt := math.Pow(hsize, plugin.Magic)
	scale := felt / hsize
	adjustedPercent := 100.0 - ((100.0 - percent) * scale)
	if plugin.Verbose {
		fmt.Printf("hsize: %v felt: %v scale %v adjusted: %v\n", hsize, felt, scale, adjustedPercent)
	}
	if adjustedPercent < 0 {
		return 0.0
	}
	if adjustedPercent > 100 {
		return 100.0
	}
	return adjustedPercent
}
